dev/AppLifecycle/AppInstance.cpp

```diff
@@ -0,0 +1,415 @@
+            {
+                auto strong_this{ weak_this.get() };
+                if (strong_this)
+                {
+                    strong_this->ProcessRedirectionRequests();
+                }
+            };
+
+            m_activationWatcher.create(m_innerActivated.get(), onInnerActivated);
+        }
+        else
+        {
+            releaseOnExit = m_dataMutex.acquire();
+        }
+
+        m_key.Open(m_processName + L"_Key");
+
+        if (m_isCurrent)
+        {
+            m_instances.Insert(processId);
+        }
+        else
+        {
+            m_instanceHandle.reset(OpenProcess(SYNCHRONIZE, FALSE, processId));
+
+            // Create a monitor thread to handle cleaning up this instance if the backing process terminates.
+            auto onInstanceTerminated = [weak_this]
+            {
+                auto strong_this{ weak_this.get() };
+                if (strong_this)
+                {
+                    strong_this->OnInstanceTerminated();
+                }
+            };
+
+            m_terminationWatcher.create(m_instanceHandle.get(), onInstanceTerminated);
+        }
+
+        m_redirectionArgs.Init(m_processName + L"_RedirectionQueue");
+    }
+
+    void AppInstance::OnInstanceTerminated()
+    {
+        auto releaseOnExit = m_dataMutex.acquire();
+        m_instances.Remove(m_processId);
+    }
+
+    GUID AppInstance::DequeueRedirectionRequestId()
+    {
+        auto releaseOnExit = m_dataMutex.acquire();
+        auto id = m_redirectionArgs.Dequeue();
+        return id;
+    }
+
+    void AppInstance::EnqueueRedirectionRequestId(GUID id)
+    {
+        auto releaseOnExit = m_dataMutex.acquire();
+        m_redirectionArgs.Enqueue(id);
+    }
+
+    void AppInstance::ProcessRedirectionRequests()
+    {
+        m_innerActivated.ResetEvent();
+
+        GUID id;
+        while ((id = DequeueRedirectionRequestId()) != GUID_NULL)
+        {
+            wil::unique_cotaskmem_string idString;
+            THROW_IF_FAILED(StringFromCLSID(id, &idString));
+
+            auto name = wil::str_printf<std::wstring>(c_requestPacketNameFormat, m_processName.c_str(), idString.get());
+
+            RedirectionRequest request;
+            request.Open(name);
+            auto args = request.UnmarshalArguments();
+
+            m_activatedEvent(*this, args);
+
+            std::wstring eventName = name + c_activatedEventNameSuffix;
+            wil::unique_event cleanupEvent;
+            cleanupEvent.open(eventName.c_str());
+            if (cleanupEvent)
+            {
+                // If the event is missing, it means the waiter gave up.  Ignore the error.
+                cleanupEvent.SetEvent();
+            }
+        }
+    }
+
+    IAsyncAction AppInstance::QueueRequest(AppLifecycle::AppActivationArguments args)
+    {
+        auto strongThis{ get_strong() };
+
+        // Push this work onto a background thread.
+        co_await resume_background();
```

Comment:
this is a threadpool thread, nothing below apears to require COM initialization, is that correct?